### PR TITLE
Refactor/enrich the rest of Quark's tests

### PR DESCRIPTION
**Pull request by haeter525**
#### Description

Please refer [here](https://github.com/quark-engine/gsoc2021-ShengFengLu/issues/1). For the replacement of Androguard, I want to write tests to improve the test coverage of Quark. This is the **final** PR. (You can find the previous PR [here](https://github.com/quark-engine/quark-engine/pull/182) )

In this PR, I focus on these files.

+ `quark/Objects/analysis.py`
+ `quark/Objects/quarkrule.py`
+ `quark/report.py`
+ `quark/freshquark.py`
+ all seven files in `quark/utils` (colors.py, graph.py, pprint.py, output.py, etc.)
#### Code Changes
+ **For the existing tests**: Divide them by their test scenarios.
+ **For the new tests**: Add them according to two strategies and the coding guideline discussed in the above issue.

| Files               | # Tests added for normal inputs | # Tests added for error inputs | # Tests modified |
| ------------------- | :-----------------------------: | :----------------------------: | :--------------: |
| `test_analysis.py`  |                1                |               0                |        0         |
| `test_quarkrule.py` |                1                |               4                |        -         |
| `test_report.py`    |                4                |               6                |        -         |
| `test_freshquark.py`   |                2                |               0                |        -         |
| `test_colors.py`    |                1                |               0                |        -         |
| `test_graph.py`     |                3                |               0                |        -         |
| `test_output.py`    |                3                |               0                |        -         |
| `test_pprint.py`    |                5                |               0                |        -         |
| `test_regex.py`     |               11                |               3                |        -         |
| `test_tools.py`     |                5                |               1                |        1         |
| `test_weight.py`    |                0                |               0                |        3         |
| **Total**           |             **36**              |             **14**             |      **4**       |

#### Related Discussions

1. issue [https://github.com/quark-engine/gsoc2021-ShengFengLu/issues/1](https://github.com/quark-engine/gsoc2021-ShengFengLu/issues/1)
2. Discussion [https://github.com/quark-engine/quark-engine/discussions/173](https://github.com/quark-engine/quark-engine/discussions/173)

### CODE REVIEW

**Comment by krnick** [comment]
Thank you @haeter525 for the excellentwork!
It seems that CI has failed the test. Could you please address it?

**Comment by haeter525** [comment]
Hi, @krnick 

It looks like CI didn't provide the package `graphviz`, which is a requirement of Quark.

I propose to add the following lines into `.travis.yml` to have it installed.
```
before_install:
  - sudo apt-get -y install graphviz
```

Please take a look at the following commit. Thanks.

File: tests/utils/test_graph.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    )
+
+
+@pytest.fixture(scope="function")
+def connect_method_2(analysis_object):
+    return next(
+        analysis_object.find_methods(
+            "Lahmyth/mine/king/ahmyth/ConnectionManager\\$1;",
+            "<init>",
+            "\\(\\)V",
```
Comment (krnick) [inline tests/utils/test_graph.py:54]: Why do descriptors `\\(\\)V` need to be escaped?

**Review by krnick — COMMENTED** [review]
Nice work!
Thank you @haeter525 for helping Quark greatly improve the test coverage and consider many aspects of the test problem.

There is only one thing I am not sure that you use `find_methods` to search method with the escape character `\` inside descriptor in file `tests/utils/test_graph.py `.

For example:

```
 analysis_object.find_methods(
		 "Ljava/lang/Object;",
		 "<init>",
		 "\\(\\)V",
```

In fact, it could be written like this:

```
 analysis_object.find_methods(
		 "Ljava/lang/Object;",
		 "<init>",
		 "()V",
```

Is there any reason?

**Comment by haeter525** [comment]
Hi @krnick, thank you for your review!

The reason is a problem with regular expressions.

`find_methods` expects regular expressions as its inputs. Therefore, the parentheses need '\\' to match the descriptor properly.

**Comment by krnick** [comment]
In fact, [`find_method`](https://github.com/quark-engine/quark-engine/blob/5fc368862969fef174146d0a525d6ec26749f00a/quark/Objects/apkinfo.py#L141-L143)(implemented by Apkinfo class) has already handled it, so there is no need to manually add escape character. I would recommand using Apkinfo(apk).find_method() to search the method.

Also, whenever you propose a new PR, please do the following command locally and make sure that every test passes.

```
pipenv --rm
pipenv install --skip-lock --dev
pytest
```

**Review by krnick — APPROVED** [review]

**Comment by haeter525** [comment]
Hi @krnick 

I've pushed a commit to change all the backslashes into the use of find_method in apkinfo.py.
Please take a look. Thank you!

**Comment by krnick** [comment]
Thanks for the patch!. LGTM

**Pull request merged**